### DISCUSSION TITLE
How to actually launch Ghidra after setting it up inside Eclipse?

### DISCUSSION THREAD

**Discussion by hippietrail** [NationalSecurityAgency/ghidra#7043] [Q&A] [answered]
I'm not a Java guy and Eclipse is about the least intuitive tool I've ever used but this is probably incredibly obvious to everyone else, which might be why it's not explained in [the instructions](https://github.com/NationalSecurityAgency/ghidra/blob/master/README.md#advanced-development):

I follow these steps just fine:

> Prepare the development environment:
> gradle prepdev eclipse buildNatives
> Import Ghidra projects into Eclipse:
> File -> Import...
> General | Existing Projects into Workspace
> Select root directory to be your downloaded or cloned ghidra source repository
> Check Search for nested projects
> Click Finish

But when I get to this step I don't know what I'm supposed to do:

> When Eclipse finishes building the projects, Ghidra can be launched and debugged with the provided Ghidra Eclipse run configuration.

There's nothing in "Run As", but there is a Ghidra icon.
<img width="379" alt="Screenshot 2024-10-12 at 6 52 13 PM" src="https://github.com/user-attachments/assets/5ead76ff-266c-49ec-9c88-3d5dbd6a90fd">

But clicking on that fails without much hint what I'm doing wrong:
<img width="577" alt="image" src="https://github.com/user-attachments/assets/11c5d141-ef74-4ab0-9ded-f3e51b993635">

Going into the configurations makes it look like several are there, so I might think it's set up but it doesn't seem to be. Am I supposed to change the "Project" or "Main class" here?
<img width="941" alt="Screenshot 2024-10-12 at 6 54 24 PM" src="https://github.com/user-attachments/assets/7b11aa1e-b32e-46a5-b1f0-3c414e7c4a7c">

Clicking on the Project's "Browse" button brings up a huge listview of... stuff? None have an obvious name like "Ghidra". Am I supposed to choose one of these?
<img width="431" alt="image" src="https://github.com/user-attachments/assets/b64886a3-94ba-4afd-9bf7-a3fa4121f82c">

Sorry but I'm totally lost.

**Comment by d-millar** [comment]
@hippietrail  From the same Import window, type “Launch” in the Search field and use that entry to import Launch Configuration’s.  “Ghidra” should be among them.

**Comment by Wall-AF** [comment]
Try the following configuration:
![image](https://github.com/user-attachments/assets/5266523b-0179-4793-ac37-0125a6d17e12)

![image](https://github.com/user-attachments/assets/d55759e4-2b85-46aa-8a08-3c0f673b241a)
The `VM arguments` I have (that are missing in the above screenshot,, sorry) are:
```
-XX:+IgnoreUnrecognizedVMOptions
-Djava.system.class.loader=ghidra.GhidraClassLoader
-Xshare:off
-Dfile.encoding=UTF8
-Duser.country=US
-Duser.language=en
-Dsun.java2d.pmoffscreen=false
-Dsun.java2d.xrender=true
-Dsun.java2d.d3d=false
-Xdock:name="Ghidra"
-Dvisualvm.display.name=Ghidra
-Dpython.console.encoding=UTF-8
--add-opens=java.desktop/sun.awt.image=ALL-UNNAMED
```

![image](https://github.com/user-attachments/assets/c587fc84-40ea-4a67-912d-2286be471ea0)

![image](https://github.com/user-attachments/assets/83470778-2496-49cd-88a1-f545c3496fb6)

**Comment by hippietrail** [comment]
Strange I'm not sure what I did that fixed it.

First I tried to follow [d-millar](https://github.com/d-millar)'s advice by going back to "File/Import..." and typing "Launch" in the "Select an import wizard field", which showed this: 
<img width="201" alt="image" src="https://github.com/user-attachments/assets/f35378a4-ffcb-4ed9-b1b8-1d01dfd62304">

But selecting it and clicking on "Next" seemed to do nothing and reach a dead end with an empty "Import Launch Configurations" browser window: 
<img width="592" alt="image" src="https://github.com/user-attachments/assets/8d649d47-388b-45e1-ae31-8811638f0623">

So then I tried to follow [Wall-AF](https://github.com/Wall-AF)'s advice but when I click on the Debug or Run icons something has changed and the menus are now populated: 
<img width="274" alt="image" src="https://github.com/user-attachments/assets/a7fa046e-5650-422c-842b-43ffe1ad751f">

And from here it "just works" and I don't have to keep following Wall-AF's guide. It all feels very nondeterministic. My best guess is that something happened silently in the background just by going to the Import Launch Configurations?

Thank you both for your help.

**Reply by Wall-AF** [reply]
IIRC, I only had to set that up once (and maybe adjust it if I had either a new JDK or something similar), then I too justt press `F11`.

**Comment by hippietrail** [comment]
After running it and quitting a couple of times to make sure it was fine, I tried renaming the function `getIcon` in `FSBIcons.java`. I did that change in VS Code before I could find the file in Eclipse. Then I found it and opened it and tried 'Run' to get it to build to see what the errors were, but instead I got what seemed like unrelated build errors not mentioning my change. Eclipse wouldn't let me save the file after undoing the change. I thought it must be busy in the background but System Monitor showed nothing was happening. I thought it's some Eclipse glitch and quit Eclipse and restarted it, picked the same workspace.

But now the Run configurations are unpopulated again and trying the same steps that mysteriously seemed to work last time now mysteriously do nothing. Even stranger, it remembered what file was open but now can't open it. 
<img width="519" alt="image" src="https://github.com/user-attachments/assets/5ac7adac-45a0-4415-be9d-7fc018c11a04">

VS Code can still open it so I know there's nothing wrong with the HD or the files.

I quit Eclipse again and restarted it. The Package Explorer shows everything is there. The "Problems" output now shows `0`. Project/Build All is greyed out. If I try to follow Wall-F's instructions I do not now have any subentries under "Java Application": 
<img width="281" alt="image" src="https://github.com/user-attachments/assets/f268ed59-9be0-40f8-b0ed-cb2999b8cb59">

**Reply by Wall-AF** [reply]
Have you selected everything in eclipse's `Package Explorer` and done a `Refresh`?

**Reply by hippietrail** [reply]
Ah that's something I didn't really know about. It did the job. It was hard to be sure because Eclipse or Gradle or Ghidra? seems to have a bug where it considers everything with a `.java` extension to be part of the project/workspace but when the project is on a FAT partition on macOS the OS often creates sidecar files with a `._` prefix and duplicating the rest of the filename including the `.java` extension but is not a text files. This generated many distracting false errors that Ghidra still ran despite of.

**Comment by d-millar** [comment, marked as answer]
Apologies - I should have been more specific, although all of @Wall-AF's comments are accurate and on-point.  Just for completeness, here is what I normally do:

(1) File->Import (brings up "Select / Choose import wizard")
(2) Either enter "Launch" in the Select field or navigate to "Run/Debug"->"Launch Configurations".  Select that, hit "Next".
(3) Navigate to "ghidra" by typing in the path or using "Browse".  Select the box next to "ghidra" in the tree below.  Hit Finish.

This should give you 8 config, accessible from the Run or Debug pull-downs:  GhidraDev, Launch Runtime Eclipse, Sleigh, PyGhidra Interpreter Debug, PyGhidra GUI Debug, Headless, JShell, and Ghidra (the last being the one I normally default to).  I think the key step you missed above is (3), browsing to ghidra's root directory.

**Reply by hippietrail** [reply]
Because of extra problems trying to hack Ghidra on a Mac from a FAT partition, I dug out another external HD that has a Mac formatted partition and set it all up again. This was definitely the bit I missed. I sometimes follow instructions to the letter when working with something unfamiliar to make sure I don't introduce problems based on my assumptions and guesses. This step was probably obvious to people who were more familiar.

**Reply by d-millar** [reply]
Well, frankly have missed more obvious things regularly - even with tools I theoretically know inside and out.  Let us know if you get stuck anywhere else!

**Reply by hippietrail** [reply]
This part should really be in the build or dev instructions. After doing a new clone of Ghidra I ended up stuck in the very same place and had to hunt through the old issues and discussions until I found this.

**Reply by d-millar** [reply]
OK, if I read the above correctly, things are working, yes?  Apologies for the lack of clarity in my instructions - I was suggesting you re-clone from "master", re-do the setup, and migrate your changes from your existing repo to the clean one, but sounds like @Wall-AF 's suggestion saved you some works.  Thanks @Wall-AF!

For future reference, most of this is covered in the [DevGuide](https://github.com/NationalSecurityAgency/ghidra/blob/master/DevGuide.md) and the plugins' [README](https://github.com/NationalSecurityAgency/ghidra/blob/master/GhidraBuild/EclipsePlugins/GhidraDev/GhidraDevPlugin/README.md).  I think they do a better job than I could of explaining the various steps, so won't reiterate them here.  I will agree gradle and git, while incredibly powerful, are not always the easiest tools to use.

**Reply by hippietrail** [reply]
No problem! Everything worked out in the end thanks to all who helped. I normally would only look at the GhidraDev docs when working on my `Loader`s and `FileSystem`s but this time I was working on my Ghidra fork.

**Comment by ryanmkurtz** [comment]
FYI: https://github.com/NationalSecurityAgency/ghidra/blob/master/DevGuide.md#eclipse-issues
